### PR TITLE
chore(controller): Refactor the CronWorkflow schedule logic with sync.Map

### PR DESCRIPTION
**Pull request by sarabala1979**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not.

### CODE REVIEW

File: workflow/cron/controller.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ type Controller struct {
 	managedNamespace     string
 	instanceId           string
 	cron                 *cron.Cron
-	nameEntryIDMap       map[string]cron.EntryID
+	nameEntryIDMap       sync.Map
 	nameEntryIDMapLock   *sync.Mutex
```
Comment (alexec) [inline workflow/cron/controller.go:44]: delete this?

File: workflow/cron/controller.go [hunk 2]
```diff
@@ -126,21 +126,19 @@ func (cc *Controller) processNextCronItem() bool {
 		return false
 	}
 	defer cc.cronWfQueue.Done(key)
-	logCtx := log.WithField("key", key)
+	logCtx := log.WithField("cronworkflow", key)
```
Comment (alexec) [inline workflow/cron/controller.go:129]: revert?

**Review by alexec — APPROVED** [review]
two minor comments - backport?

File: workflow/cron/controller.go [hunk 3]
```diff
@@ -126,21 +126,19 @@ func (cc *Controller) processNextCronItem() bool {
...
 	logCtx.Infof("Processing %s", key)
 
 	obj, exists, err := cc.cronWfInformer.Informer().GetIndexer().GetByKey(key.(string))
 	if err != nil {
 		logCtx.WithError(err).Error(fmt.Sprintf("Failed to get CronWorkflow '%s' from informer index", key))
 		return true
 	}
-	cc.nameEntryIDMapLock.Lock()
```
Comment (sarabala1979) [inline workflow/cron/controller.go:137]: Which is block entire block for all the workers to schedule the CronWorkflows.

File: workflow/cron/controller.go (hunk 2 above)
Comment (sarabala1979) [reply workflow/cron/controller.go:129]: We need to distinct  the CronWorkflow controller logs from workflow controller. This `key` is Cronworkflow Name.

File: workflow/cron/controller.go (hunk 1 above)
Comment (sarabala1979) [reply workflow/cron/controller.go:44]: good catch

File: workflow/cron/controller.go (hunk 2 above)
Comment (alexec) [reply workflow/cron/controller.go:127]: ok, can it be `cronWorkflow`?

Comment (alexec) [reply workflow/cron/controller.go:127]: capital W

File: workflow/cron/controller.go [hunk 4]
```diff
@@ -124,7 +124,7 @@ func (cc *Controller) processNextCronItem() bool {
 		return false
 	}
 	defer cc.cronWfQueue.Done(key)
-	logCtx := log.WithField("cronworkflow", key)
+	logCtx := log.WithField("CronWorkflow", key)
```
Comment (alexec) [inline workflow/cron/controller.go:127]: lower case 'c'!

Comment (sarabala1979) [reply workflow/cron/controller.go]: Sorry by mistake

**Review by alexec — APPROVED** [review]

**Pull request merged**